### PR TITLE
Er 818 status product refactoring

### PR DESCRIPTION
**Pull request by wiewiurdp**
# Description

Status product refactoring

Issue #1218 
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [ ] Unit Tests
- [ ] Behat Test

# Checklist:
- [x] I have read the contribution requirements and fulfill them.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: module/product/src/Infrastructure/Strategy/ProductStrategyInterface.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+declare(strict_types=1);
+
+namespace Ergonode\Product\Infrastructure\Strategy;
+
+use Ergonode\Product\Domain\Entity\AbstractProduct;
+use Ergonode\Product\Domain\ValueObject\Sku;
+use Ergonode\SharedKernel\Domain\Aggregate\ProductId;
+use Ergonode\SharedKernel\Domain\Aggregate\TemplateId;
+
+interface ProductStrategyInterface
```
Comment (rprzedzik) [inline module/product/src/Infrastructure/Strategy/ProductStrategyInterface.php:17]: The class name does not describe its purpose.

File: module/product/src/Infrastructure/Strategy/SimpleProductStrategy.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+namespace Ergonode\Product\Infrastructure\Strategy;
+
+use Ergonode\Product\Domain\Entity\AbstractProduct;
+use Ergonode\Product\Domain\Entity\SimpleProduct;
+use Ergonode\Product\Domain\ValueObject\Sku;
+use Ergonode\SharedKernel\Domain\Aggregate\ProductId;
+use Ergonode\SharedKernel\Domain\Aggregate\TemplateId;
+
+class SimpleProductStrategy implements ProductStrategyInterface
```
Comment (rprzedzik) [inline module/product/src/Infrastructure/Strategy/SimpleProductStrategy.php:18]: The class name does not describe its purpose.

File: module/product/src/Infrastructure/Strategy/VariableProductStrategy.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace Ergonode\Product\Infrastructure\Strategy;
+
+use Ergonode\Product\Domain\Entity\AbstractProduct;
+use Ergonode\Product\Domain\Entity\VariableProduct;
+use Ergonode\Product\Domain\ValueObject\Sku;
+use Ergonode\SharedKernel\Domain\Aggregate\ProductId;
+use Ergonode\SharedKernel\Domain\Aggregate\TemplateId;
+
+class VariableProductStrategy implements ProductStrategyInterface
```
Comment (rprzedzik) [inline module/product/src/Infrastructure/Strategy/VariableProductStrategy.php:18]: The class name does not describe its purpose.

File: module/product/src/Resources/config/services.yml [hunk 4]
```diff
@@ -57,3 +57,7 @@ services:
         decorates: Ergonode\Product\Domain\Query\ProductQueryInterface
         arguments:
             - '@Ergonode\Product\Infrastructure\Persistence\Query\Decorator\CacheProductQueryDecorator.inner'
+
+    Ergonode\Product\Infrastructure\Provider\ProductStrategyProvider:
+        arguments:
+            $productStrategies: !tagged_iterator 'component.product.product_strategy'
```
Comment (rprzedzik) [inline module/product/src/Resources/config/services.yml:63]: tag name does not describe its purpose.

File: module/workflow/src/Domain/Factory/ProductFactoryDecorator.php [hunk 5]
```diff
@@ -1,52 +1,58 @@
...
+use Ergonode\Product\Domain\Factory\ProductFactoryInterface;
+use Ergonode\Product\Domain\ValueObject\Sku;
+use Ergonode\SharedKernel\Domain\Aggregate\ProductId;
+use Ergonode\SharedKernel\Domain\Aggregate\TemplateId;
 use Ergonode\Value\Domain\ValueObject\TranslatableStringValue;
+use Ergonode\Workflow\Domain\Entity\Attribute\StatusSystemAttribute;
+use Ergonode\Workflow\Domain\Provider\WorkflowProvider;
 
-abstract class AbstractCreateProductHandler
+class ProductFactoryDecorator implements ProductFactoryInterface
```
Comment (rprzedzik) [inline module/workflow/src/Domain/Factory/ProductFactoryDecorator.php:24]: The class name does not describe its purpose.

File: module/product/src/Infrastructure/Strategy/ProductStrategyInterface.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+{
+    public function supports(string $type): bool;
+
+    public function build(
+        ProductId $id,
+        Sku $sku,
+        TemplateId $templateId,
+        array $categories,
+        array $attributes
+    ): AbstractProduct;
```
Comment (rprzedzik) [inline module/product/src/Infrastructure/Strategy/ProductStrategyInterface.php:27]: IMHO  this inteface should be open for any potential informations that be needed for catering any product Class

File: module/product/src/Domain/Factory/ProductFactory.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+     */
+    private function addAudit(array $attributes): array
+    {
+        $user = $this->security->getUser();
+        if ($user) {
+            $value = new StringValue(sprintf('%s %s', $user->getFirstName(), $user->getLastName()));
+            $attributes[CreatedBySystemAttribute::CODE] = $value;
+        }
+
+        return $attributes;
```
Comment (rprzedzik) [inline module/product/src/Domain/Factory/ProductFactory.php:59]: what about CreateAtSystemAttribute

**Review by rprzedzik — CHANGES_REQUESTED** [review]
Also i thing this task required migration to fix missing product statuses.

File: module/product/src/Infrastructure/Strategy/ProductStrategyInterface.php (hunk 6 above)
Comment (piotrkreft) [reply module/product/src/Infrastructure/Strategy/ProductStrategyInterface.php:27]: Maybe a second interface can be created extending this one with an additional optional parameter.

**Review by piotrkreft — APPROVED** [review]

**Review by rprzedzik — APPROVED** [review]

**Pull request merged**